wordpress-mobile/WordPress-Android
# Giphy Picker

This completes the Giphy Picker Feature #8627.

## Merged Pull Requests

The work was broken down into multiple pull requests which were reviewed by @kwonye.

- shiki/WordPress-Android#1 – First Iteration. Searching and endless scroll
- shiki/WordPress-Android#2 – Item selection
- shiki/WordPress-Android#3 – Importing in Media Library
- shiki/WordPress-Android#4 – Preview
- shiki/WordPress-Android#5 – Empty results view
- shiki/WordPress-Android#6 – Importing in Post Editor
- shiki/WordPress-Android#7 – Error handling
- shiki/WordPress-Android#8 – Searching while typing
- shiki/WordPress-Android#9 – Attribution in licenses.html

## Features

The Stock Media Picker was the main inspiration for how this Giphy Picker works.

### Searching

![dec-14-2018 20-44-55](https://user-images.githubusercontent.com/198826/50038724-509d2300-ffe1-11e8-947d-e4b2299a53ba.gif)

Search happens as the user types. The result list shows an accurate progress bar because the Giphy API responds with the total number of GIFs we can expect. Subsequent pages are loaded as the user scrolls.

### Selection

![dec-14-2018 20-48-36](https://user-images.githubusercontent.com/198826/50038751-aa9de880-ffe1-11e8-8275-df139bcf9843.gif)

Users can select multiple items. The _Add_ and _Preview_ buttons at the bottom reflect how many items were selected.

### Importing

![dec-14-2018 20-50-04](https://user-images.githubusercontent.com/198826/50038756-de790e00-ffe1-11e8-8fa4-a6b58f06312f.gif)

![dec-14-2018 20-53-03](https://user-images.githubusercontent.com/198826/50038775-462f5900-ffe2-11e8-87ca-8c8e01d2312b.gif)

Importing can be done from the Media Library or the Post Editor. It generally works in 2 steps:

1. The selected GIFs are downloaded concurrently from Giphy.
2. The downloaded files are passed to the Media Library UI or the Post Editor which uploads them to the Media Library.

The _Add_ button is replaced with a circular progress bar when downloading. This is based on the [Integrating with actions section](https://material.io/design/components/progress-indicators.html#circular-progress-indicators) in Material Design.

### Preview

![dec-14-2018 20-55-24](https://user-images.githubusercontent.com/198826/50038787-a4f4d280-ffe2-11e8-8571-c20bd5335b69.gif)

The preview uses the existing Media Preview component. Multiple items can be previewed. Performing a long press on a single item will also preview it.

## Testing

1. Create a [Giphy API key](https://developers.giphy.com/docs/) and paste it in your `gradle.properties` as the value for `wp.giphy.api_key`
2. Go to the Media Library
3. Click on the _Add_ button at the top right. Select the _Choose from Giphy_ option
4. In the Giphy picker, search and pick some GIFs
5. Press the _Add_ button. The selected GIFs should be downloaded and the picker should be closed. The upload will commence in the Media Library.

### Possible Regressions

Some methods used by other parts of the app were changed. These may have regressions:

- Places that use Photo Picker like _Me_ or the _Media Library_
- Stock Media Picker

### Known Issues

If you test this long enough, you will eventually hit API rate limits. This will cause empty blocks in the grid.

## Production Configuration

A new `gradle.properties` setting is required for this to work – `wp.giphy.api_key`.
